File: matlab/ex2.m
%   ex2.m
%   Copyright (C) 2020  Zalan Rajna
%
%   This program is free software: you can redistribute it and/or modify
%   it under the terms of the GNU General Public License as published by
%   the Free Software Foundation, either version 3 of the License, or
%   (at your option) any later version.
%
%   This program is distributed in the hope that it will be useful,
%   but WITHOUT ANY WARRANTY; without even the implied warranty of
%   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%   GNU General Public License for more details.
%
%   You should have received a copy of the GNU General Public License
%   along with this program.  If not, see <https://www.gnu.org/licenses/>.

close all;
clear;
tic;

%% Task 1
%   Region growing segmentation
%   Hint: you should/can use imsegfmm function in fast_march

abdomen = imread(fullfile('images','abdomen.png'));

figure;

subplot(121);
imshow(abdomen);
title('original CT section');

seedpoint = [330, 170];

hold on;
plot(seedpoint(1), seedpoint(2), 'r+', 'MarkerSize', 30, 'LineWidth', 2);
legend('seed point', 'Location', 'bestoutside');

lung = fast_march(abdomen, seedpoint);

lung_overlaid = imoverlay(abdomen, lung, 'green');

subplot(122);
imshow(lung_overlaid);
title('segmented lung');

%% Task 2
%   Active contours segmentation
%   Hint: you should/can use activecontour function in segment_cells

cells = imread(fullfile('images','cells.png'));
cells = normalize_image(cells);

figure;

subplot(121);
imshow(cells);
title('original cells and initial mask');

roi = images.roi.Circle(gca, 'Center', [460, 430], 'Radius', 300);
mask = roi.createMask();

bin_result = segment_cells(cells, mask);

cells_overlay = imoverlay(cells, bin_result, 'green');

subplot(122);
imshow(cells_overlay);
title('segmented cells')

%% Task 3
%   Watershed segmentation
%   Hint: you should/can use watershed function in segment_colon

colon = imread(fullfile('images','colon.tif'));

figure;

subplot(121);
imshow(colon);
title('original colon image');

colon = rgb2gray(colon);
skeleton = segment_colon(colon);

colon_overlay = imoverlay(colon, skeleton, 'green');

subplot(122);
imshow(colon_overlay);
title('segmented colon (skeleton lines)')

%% Runtime

disp(['Runtime (in seconds) was: ', num2str(toc)]);

%% Helper functions
%   You should write your functions in separate files

function J = normalize_image(I)
% Normalizes the image to have values stretched between 0 and 1
    J = im2double(I);
    J = J - min(J(:));
    J = J ./ max(J(:));
end

File: python/ex2.py
#   ex2.py
#   Copyright (C) 2020  Zalan Rajna
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <https://www.gnu.org/licenses/>.

from os import path

import numpy as np
import cv2.cv2 as cv
from matplotlib import pyplot as plt
from skimage.draw import circle_perimeter

from ex2_functions import segment_lung, segment_cells, segment_colon


def task1():
    """Task 1: Region growing segmentation"""
    plt.figure()
    abdomen = cv.imread(path.join("images", "abdomen.png"), cv.IMREAD_GRAYSCALE)

    plt.subplot(1, 2, 1)
    plt.imshow(abdomen, cmap="gray")
    plt.axis("off")
    plt.title("original CT section")

    seedpoint = (330, 170)

    plt.plot(seedpoint[0], seedpoint[1], color='red', marker="+", markersize=20, linewidth=2)
    plt.legend('seed point', loc='upper left')

    lung = segment_lung(abdomen, seedpoint)

    background = np.zeros(abdomen.shape + (3, ), dtype=np.uint8)
    background[..., 1] = 255
    abdomen_masked = np.ma.masked_where(lung, abdomen)
    plt.subplot(1, 2, 2)
    plt.imshow(background)
    plt.imshow(abdomen_masked, cmap="gray")
    plt.axis("off")
    plt.title("segmented lung")

    plt.show()


def task2():
    """Task 2: Active contours segmentation"""
    plt.figure()
    cells = cv.imread(path.join("images", "cells.png"), cv.IMREAD_GRAYSCALE)

    plt.subplot(1, 2, 1)
    plt.imshow(cells, cmap="gray")
    plt.axis("off")
    plt.title("original cells and initial mask")

    center = (460, 430)
    radius = 300
    circle = plt.Circle(center, radius, color='deepskyblue', alpha=0.2)
    ax = plt.gca()
    ax.add_artist(circle)
    mask = np.zeros(cells.shape, dtype=np.uint8)
    mask[circle_perimeter(center[0], center[1], radius)] = True

    cells_bin = segment_cells(cells, mask)

    background = np.zeros(cells.shape + (3, ), dtype=np.uint8)
    background[..., 1] = 255
    abdomen_masked = np.ma.masked_where(cells_bin, cells)
    plt.subplot(1, 2, 2)
    plt.imshow(background)
    plt.imshow(abdomen_masked, cmap="gray")
    plt.axis("off")
    plt.title("segmented cells")

    plt.show()


def task3():
    """Task 3: Watershed segmentation"""
    plt.figure()
    colon = cv.imread(path.join("images", "colon.tif"))
    colon = cv.cvtColor(colon, cv.COLOR_BGR2RGB)

    plt.subplot(1, 2, 1)
    plt.imshow(colon)
    plt.axis("off")
    plt.title("original colon image")

    skeleton = segment_colon(colon)

    background = np.zeros(colon.shape, dtype=np.uint8)
    background[..., 1] = 255
    colon_gray = cv.cvtColor(colon, cv.COLOR_RGB2GRAY)
    colon_masked = np.ma.masked_where(skeleton, colon_gray)
    plt.subplot(1, 2, 2)
    plt.imshow(background)
    plt.imshow(colon_masked, cmap="gray")
    plt.axis("off")
    plt.title("segmented cells")

    plt.show()


def main():
    """ICVMBI exercise 2"""
    task1()
    task2()
    task3()


if __name__ == "__main__":
    main()
